### PR TITLE
UCT/TEST: Fix deadlock with dlopen() from multiple threads

### LINKED ISSUE #3517: Deadlock in ucm_dlopen
```

Thread 3 (Thread 0x7f4ff7133700 (LWP 38421)):
#0  0x00007f501890042d in __lll_lock_wait () from /usr/lib64/libpthread.so.0
#1  0x00007f50188fbde6 in _L_lock_870 () from /usr/lib64/libpthread.so.0
#2  0x00007f50188fbcdf in pthread_mutex_lock () from /usr/lib64/libpthread.so.0
#3  0x00007f5018662632 in _dl_addr () from /usr/lib64/libc.so.6
#4  0x00007f50099becb2 in ucm_reloc_modify_got (phname=<optimized out>, ctx=0x7f4ff7132cb0, ctx=0x7f4ff7132cb0, phsize=<optimized out>, phnum=<optimized out>, phdr=<optimized out>, base=139981991448576)
    at util/reloc.c:197
#5  ucm_reloc_phdr_iterator (info=<optimized out>, size=<optimized out>, data=0x7f4ff7132cb0) at util/reloc.c:241
#6  0x00007f501866250c in dl_iterate_phdr () from /usr/lib64/libc.so.6
#7  0x00007f50099bf030 in ucm_reloc_apply_patch (patch=0x7f5009bc9ce0 <ucm_reloc_dlopen_patch>) at util/reloc.c:269
#8  ucm_dlopen (filename=<optimized out>, flag=<optimized out>) at util/reloc.c:299
#9  0x00007f5018e1bed1 in sharp_coll_cuda_context_init (context=0x7f4fdc0008c0) at cuda_util.c:369
#10 0x00007f5018e0e2dc in sharp_coll_init (sharp_coll_spec=0x7f4ff7132e50, sharp_coll_context=0x7f4ff71336f0) at context.c:307
#11 0x0000000000401489 in test_comm_create (arg=0x7ffea432ea60) at examples/thread.c:105
#12 0x00007f50188f9e25 in start_thread () from /usr/lib64/libpthread.so.0
#13 0x00007f501862734d in clone () from /usr/lib64/libc.so.6

Thread 2 (Thread 0x7f4ff6932700 (LWP 38422)):
#0  0x00007f501890042d in __lll_lock_wait () from /usr/lib64/libpthread.so.0
#1  0x00007f50188fbde6 in _L_lock_870 () from /usr/lib64/libpthread.so.0
#2  0x00007f50188fbcdf in pthread_mutex_lock () from /usr/lib64/libpthread.so.0
#3  0x00007f5019257fc9 in _dl_add_to_namespace_list () from /lib64/ld-linux-x86-64.so.2
#4  0x00007f5019253a5c in _dl_map_object_from_fd () from /lib64/ld-linux-x86-64.so.2
#5  0x00007f5019255588 in _dl_map_object () from /lib64/ld-linux-x86-64.so.2
#6  0x00007f50192609a4 in dl_open_worker () from /lib64/ld-linux-x86-64.so.2
#7  0x00007f501925c314 in _dl_catch_error () from /lib64/ld-linux-x86-64.so.2
#8  0x00007f501926030b in _dl_open () from /lib64/ld-linux-x86-64.so.2
#9  0x00007f501832bfbb in dlopen_doit () from /usr/lib64/libdl.so.2
#10 0x00007f501925c314 in _dl_catch_error () from /lib64/ld-linux-x86-64.so.2
#11 0x00007f501832c5bd in _dlerror_run () from /usr/lib64/libdl.so.2
#12 0x00007f501832c051 in dlopen@@GLIBC_2.2.5 () from /usr/lib64/libdl.so.2
#13 0x00007f50099befce in ucm_dlopen (filename=<optimized out>, flag=<optimized out>) at util/reloc.c:287
#14 0x00007f5018e1c341 in sharp_coll_cuda_context_init (context=0x7f4fd40008c0) at cuda_util.c:457
#15 0x00007f5018e0e2dc in sharp_coll_init (sharp_coll_spec=0x7f4ff6931e50, sharp_coll_context=0x7f4ff69326f0) at context.c:307
#16 0x0000000000401489 in test_comm_create (arg=0x7ffea432ea68) at examples/thread.c:105
#17 0x00007f50188f9e25 in start_thread () from /usr/lib64/libpthread.so.0
---Type <return> to continue, or q <return> to quit---
```

### PR DESCRIPTION
**Pull request by yosefe**
Fixes #3517 
@bureddy

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/10151/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/10156/ for details (Mellanox internal link).

File: src/ucm/util/reloc.c [hunk 1]
```diff
@@ -236,33 +235,34 @@ ucm_reloc_modify_got(ElfW(Addr) base, const ElfW(Phdr) *phdr, const char *phname
...
     ucm_reloc_dl_iter_context_t *ctx = data;
     int phsize;
     int i;
 
+    /* check if need to update only the one module which is being loaded */
+    if (ctx->dlopened_base_addr && (ctx->dlopened_base_addr != info->dlpi_addr)) {
+        ucm_trace("not updating symbols in '%s'", info->dlpi_name);
+        return 0;
+    }
+
```
Comment (bureddy) [inline src/ucm/util/reloc.c]: do we still need to continue at line:282 as the only module patching is done?

**Comment by yosefe** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/10169/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/10175/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
@brminich @bureddy can you pls take a look?

**Pull request merged**